lib/lcevc/lcevc_dec.js

```diff
@@ -0,0 +1,218 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2016 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+goog.provide('shaka.lcevc.Dec');
+goog.require('shaka.log');
+goog.require('shaka.Deprecate');
+goog.require('shaka.util.IReleasable');
+
+/**
+ * @summary
+ *  lcevcDec - (MPEG-5 Part 2 LCEVC - Decoder) provides
+ *  all the operations related to the enhancement and rendering
+ *  of LCEVC enabled streams and on to a canvas.
+ * @implements {shaka.util.IReleasable}
+ * @export
+ */
+shaka.lcevc.Dec = class {
+  /**
+   * @param {HTMLVideoElement} media The video element that will be attached to
+   * LCEVC Decoder for input.
+   * @param {HTMLCanvasElement} canvas The canvas element that will be attached
+   * to LCEVC Decoder to render the enhanced frames.
+   * @param {shaka.extern.LcevcConfiguration} lcevcConfig LCEVC configuration
+   * object to initialize the LCEVC Decoder.
+   */
+  constructor(media, canvas, lcevcConfig) {
+    /** @private {?} */
```

Comment:
I had tried to put the type before, but for some reason JS compiler does not accept `LCEVCDec` as an acceptable type.

Reply:
@joeyparrish ping!

Reply:
I will try to figure out the right syntax, push a commit to your fork if I find a fix, then approve and merge the PR

Reply:
I know the Closure compiler sucks and nobody knows how to work within its system, so I don't blame anybody for not figuring this out.  I'm ten-years deep in it by now, so I have seen some things.

I switched from `LCEVCdec` to `typeof LCEVCdec` (the interface to a module) and got the error `Missing type for typeof value. The value must be declared and const.`  So I went to your extern and marked the namespace with `/** @const */` like the others in that folder.

That got me to the place I said you would get, where you need a cast for `this.lcevcLib_.LcevcDil`.  I'll have a complete fix posted this afternoon (US/Pacific).

Reply:
`ERROR: Permission to v-novaltd/shaka-player.git denied to joeyparrish.`

@v-nova-vinod, when you send PRs, please consider using the `Allow edits from maintainers` feature to allow us to help you out.

In the meantime, here is the diff:

```diff
diff --git a/externs/lcevc.js b/externs/lcevc.js
index b884832ef..3134d3b87 100644
--- a/externs/lcevc.js
+++ b/externs/lcevc.js
@@ -6,7 +6,9 @@
 */
 
 // This empty namespace is declared to check if LcevcDec libraries are loaded.
+/** @const */
 var libDPIModule = {};
+/** @const */
 var LCEVCdec = {};
 
 /**
@@ -61,6 +63,13 @@ LCEVCdec.LCEVCdec = class {
   close() {}
 };
 
+/**
+ * The older module interface, for backward compatibility.
+ * Typed to the same interface, but under a different name.
+ * @type {typeof LCEVCdec.LCEVCdec}
+ */
+LCEVCdec.LcevcDil;
+
 /**
  * LCEVC Support Check
  */
diff --git a/lib/lcevc/lcevc_dec.js b/lib/lcevc/lcevc_dec.js
index ec4df6489..09aac2a8e 100644
--- a/lib/lcevc/lcevc_dec.js
+++ b/lib/lcevc/lcevc_dec.js
@@ -27,10 +27,12 @@ shaka.lcevc.Dec = class {
    * object to initialize the LCEVC Decoder.
    */
   constructor(media, canvas, lcevcConfig) {
-    /** @private {?} */
-    // LCEVC Decoder library based on availability, to check if either
-    // lcevc_dil or lcevc_dec is present.
-    this.lcevcLib_ = null;
+    /**
+     * LCEVC Decoder library based on availability, to check if either
+     * lcevc_dil or lcevc_dec is present.
+     * @private {typeof LCEVCdec}
+     */
+    this.lcevcLib_;
 
     /** @private {?LCEVCdec.LCEVCdec} */
     this.dec_ = null;
```

@v-nova-vinod, the reason this typing is important is because the compiler is able to catch lots of basic mistakes for us.  Turning it off for all calls into your module is a big gap in static analysis and compiler type safety.  Thanks for being patient and getting this fixed.  When these changes are applied, I would be happy to merge the PR.  Or @avelad can do it, since I'll be on vacation for the next week.

Thanks!

Reply:
Thank you for the explanation @joeyparrish , this is very helpful and thanks for the changes. I have made the changes and pushed. I will make sure for future PRs i will enable the option for the maintainer edits. Thanks again!!

Reply:
@v-nova-vinod There are some errors:
```
ERROR: Unable to parse a tag's type expression for source file /home/runner/work/shaka-player/shaka-player/lib/lcevc/lcevc_dec.js in line 30 with tag title "private" and text "{typeof LCEVCdec}": Invalid type expression "typeof LCEVCdec": Expected "|" but "L" found.
ERROR: Unable to parse a tag's type expression for source file /home/runner/work/shaka-player/shaka-player/lib/lcevc/lcevc_dec.js in line 35 with tag title "private" and text "{typeof LCEVCdec}": Invalid type expression "typeof LCEVCdec": Expected "|" but "L" found.
```

Reply:
There seems to be some issue with the build script that is flagging a type issue.

Reply:
I think i know what might be happening here. With @joeyparrish 's fixes on the type as `typeof`,  the build process is successful but the when the docs are generated with `jsdoc`, it errors when it finds a space in the type name. 
I see that there is a custom version of `jsdoc` used so i tried updating the jsdoc version and it gave the error once and then there was no error seen. So we might need to take a call on how to address this issue.

Reply:
If this is run twice once the docs are generated when it errors, it does not flag this error. It is just the first time post some changes to the code. So i think it could be a false alarm from the docs generation part.

Reply:
@joeyparrish ping!

Reply:
Hi @joeyparrish @avelad can you please suggest the next step on this? Thanks.

Reply:
Looking into it now.  Sorry for the delay.

Reply:
I am able to fix it by moving the `typeof` expression into a typedef in the externs.

However, permission was denied to me pushing to your fork because you are not allowing maintainer edits.

Please incorporate this diff:

```diff
diff --git a/externs/lcevc.js b/externs/lcevc.js
index 3134d3b87..c816df2a3 100644
--- a/externs/lcevc.js
+++ b/externs/lcevc.js
@@ -95,3 +95,11 @@ LCEVCdec.SupportObject.SupportStatus;
  * @type {string}
  */
 LCEVCdec.SupportObject.SupportError;
+
+/**
+ * Typedef for the module interface.  Both LCEVCdec (new module) and LcevcDil
+ * (old module) implement roughly the same interface.
+ *
+ * @typedef {typeof LCEVCdec}
+ */
+var LCEVCmodule;
diff --git a/lib/lcevc/lcevc_dec.js b/lib/lcevc/lcevc_dec.js
index 09aac2a8e..f28d3a324 100644
--- a/lib/lcevc/lcevc_dec.js
+++ b/lib/lcevc/lcevc_dec.js
@@ -30,7 +30,7 @@ shaka.lcevc.Dec = class {
     /**
      * LCEVC Decoder library based on availability, to check if either
      * lcevc_dil or lcevc_dec is present.
-     * @private {typeof LCEVCdec}
+     * @private {LCEVCmodule}
      */
     this.lcevcLib_;
 
```

Reply:
Thanks @joeyparrish!  That fixes it for me as well. I have made these changes and pushed.
I am not able to find the option to allow maintainers to allow commits on this PR. I think i will have to enable that when the PR is raised. I will keep this in mind in the next PR.